src/modules/firewall/src/lib/Firewall.h

```diff
@@ -39,32 +40,265 @@ class FirewallLog
     static OSCONFIG_LOG_HANDLE m_logHandle;
 };
 
+class StringEnum
+{
+public:
+    StringEnum() = default;
+    StringEnum(const std::string& value) : m_value(value) {}
+    virtual ~StringEnum() = default;
+
+    virtual bool IsValid() const = 0;
+
+    std::string ToString() const
+    {
+        return m_value;
+    }
+
+    bool operator==(const char* other) const
+    {
+        return m_value.compare(other) == 0;
+    }
+
+    bool operator!=(const char* other) const
+    {
+        return m_value.compare(other) != 0;
+    }
+
+    friend std::ostream& operator<<(std::ostream& os, const StringEnum& se)
+    {
+        os << se.ToString();
+        return os;
+    }
+
+protected:
+    std::string m_value;
+};
+
+class GenericRule
+{
+public:
+    class State : public StringEnum
+    {
+    public:
+        static const std::array<std::string, 2> m_values;
+
+        State() = default;
+        State(const std::string& value) : StringEnum(value) {}
+
+        bool IsValid() const override
+        {
+            return std::find(m_values.begin(), m_values.end(), m_value) != m_values.end();
+        }
+    };
+
+    class Action : public StringEnum
+    {
+    public:
+        static const std::array<std::string, 3> m_values;
+
+        Action() = default;
+        Action(const std::string& value) : StringEnum(value) {}
+        virtual ~Action() = default;
+
+        virtual bool IsValid() const override
+        {
+            return std::find(m_values.begin(), m_values.end(), m_value) != m_values.end();
+        }
+    };
+
+    class Direction : public StringEnum
+    {
+    public:
+        static const std::array<std::string, 2> m_values;
+
+        Direction() = default;
+        Direction(const std::string& value) : StringEnum(value) {}
+
+        bool IsValid() const override
```

Comment:
>bool IsValid() const override [](http://example.com/codeflow?start=8&length=29)

Should there be error logging in this code to explain failures for MmiGet and MmiSet?

Reply:
Depending on where the string enum is used (parsing, serializing, etc) `IsValid()` is used similarly to `HasParseError()`